### ISSUE TITLE
(rare/infrequent) Scrollbars disappear

### ISSUE THREAD

**Issue by Robinson7D** [DataTables/DataTables#647]
For certain tables with lots of data - many rows/columns, I'm occasionally experiencing an issue where the scrollbars (both vertical, and horizontal) disappear. I haven't created a test page yet, but I will aim to do that in the next few days.

I have, however, isolated the problem in the source code. There appears to be an off-by-one error in step 3 ("Apply the measurements") of the `_fnScrollDraw` function. While I cannot show screenshots of the actual table, I've dumped some of the data about the table to the console:

![table_off_by_one](https://cloud.githubusercontent.com/assets/1261701/9479115/c4a72480-4b49-11e5-8056-1347e849d4b0.png)

As you can see, it sees that `table.outerWidth() < sanityWidth` (due to the off-by-one), and so sets a number to the `correction` variable (which has usually had 100% to allow scrolling). This the horizontal scrollbar is deemed unnecessary, and the vertical one is moved off-screen.

I've found two potential fixes for this:

1) In CSS, a max-width can be applied:
```
.dataTables_scrollBody, .dataTables_scrollHead, .dataTables_scrollFoot, .DTFC_LeftBodyWrapper, .DTFC_LeftHeadWrapper, .DTFC_LeftFootWrapper {
    max-width: 100%;
  }
```

Alternatively:
2) In the JS, it can check if the container is smaller before calculating the correction:
```
// Horizontal scroll:
if(sanityWidth > table.parent().innerWidth()){
    correction = '100%';
}
else {
    // The min width depends upon if we have a vertical scrollbar visible or not */
    correction = ((divBodyEl.scrollHeight > divBodyEl.offsetHeight ||
        divBody.css('overflow-y') == "scroll")) ?
            sanityWidth+barWidth :
            sanityWidth;
}
```

(Currently it, the old version of the function is just the stuff in the `else` section.)

Generally, I'd prefer the CSS solution, but I'm not sure if the JS code might cause other problems.

Anyway, I will attempt to create a http://live.datatables.net/ example demonstrating this.

**Comment by DataTables** [comment]
Thanks for posting this up and providing details about it. I wonder if there is a sub-pixel rounding issue here and DataTables should possible use `getBoundingClientRect` rather than $().width() and friends. I've noticed before that browsers can sometimes round down causing similar issues.

If you are able to provide a test case, that would be super!

**Comment by Robinson7D** [comment]
Hey, sorry for taking a while to get back to you. I took yesterday as vacation, and tried to stay away from code (as much as I could). 

I'm not sure if I'll have time to create the test-case in the next few days, but I'll try to find time when I can. 

I'll definitely try plugging `getBoundingClientRect().width` in and see if it shows some rounding issues / a good solution.

**Comment by DataTables** [comment]
One day doesn't count as a while! I've got stuff on my to do list from 2014... I'm going to be traveling next week, so I won't be around much then, but will look at anything you can cook up after that if you get a chance.

**Comment by Robinson7D** [comment]
Cool, sounds good. 

I think I found out what's happening. I'm very sure, right now, that the footer width is causing this. I placed `console.log` statements throughout the function and isolated it to being in the following block, where the 1px is lost:

```
if ( footer )
{
    _fnApplyToChildren( function(nSizer, i) {
        nSizer.innerHTML = "";
        nSizer.style.width = footerWidths[i];
    }, footerSrcEls );
}
```

It appears that the header equivalent is storing the values, and placing them in as follows:
```
nSizer.innerHTML = '<div class="dataTables_sizing" style="height:0;overflow:hidden;">'+headerContent[i]+'</div>';
```

If I mimic this with the footer (storing the `footerContent` similarly, etc.) it seems to fix the issue entirely.

That said, removing the line `nSizer.innerHTML = "";` from the footer variation seems to fix it as well (and doesn't seem to cause different footer sizings, with or without data in the footer). 

So, basically, it seems like there are two additional solutions (in addition to the possible ones outlined at the start of the ticket):

1) Track footerContent in a way consistent with headerContent,

OR:

2) Avoid blanking out the footer content when setting the sizer widths.

Given the comments above these blocks, I expect that option 1) may be preferred?

```
		// "Hide" the header and footer that we used for the sizing. We need to keep
		// the content of the cell so that the width applied to the header and body
		// both match, but we want to hide it completely. We want to also fix their
		// width to what they currently are
```

**Comment by pmcbrien** [comment]
Thanks!
